segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -183,7 +183,10 @@ protected void errorHandler(Throwable ex) {
             // Shutdown exceptions means we are already stopping, so no need to do anything else. For all other cases,
             // record the failure and then stop the OperationProcessor.
             super.errorHandler(ex);
-            stopAsync();
+
+            // Run async to prevent deadlocks. closeQueue() above is the most important thing when shutting down as it
+            // will prevent anything new from being added while also cancelling in-flight requests.
+            this.executor.execute(this::stopAsync);
```

Comment:
I'm missing something silly here, isn't `stopAsync` asynchronous already?

Reply:
StopAsync initiates an async stop, but the initiation is not async. Furthermore, `stopAsync` acquires a lock (synchronously) which may or may not be already acquired. Normally it is not. However in test cases, if a test fails in a certain way, this lock is acquired by a thread (that caused the failure), then some Future that was also waiting on that operation is failed; this causes the test itself to fail (because it was waiting for the result of that future) which causes the `@Cleanup` to be invoked (i.e., `AutoCloseable.close`) for all the services involved. This causes the `stopAsync` to be invoked from a different thread (`close` must invoke `stopAsync` and await its termination), which causes a deadlock (since `stopAsync` is going to try to acquire the lock that was already acquired - on a different thread).

(See my comment above about why completing a future while holding a lock can cause deadlocks). 

I don't see a way for this to happen in production code. In production code, we have a clear path on how services are shut down in case of failure, and there is no "cycle" that causes an external actor (such as the JUnit test) to try to shut it down again.

This was the cause for several different test timeouts that are linked to this PR.

Reply:
I'm wondering if we could have instead refactored `doStop` to execute its logic asynchronously.